rfcs/convergence/native-element-props.md

```diff
@@ -117,41 +119,46 @@ Examples include:
 The `ComponentProps` type has a template parameter to specify the primary slot:
 
 ```ts
-export type ComponentProps<Shorthands, Primary extends keyof Shorthands = 'root'> =
-  // Put the primary slot's props directly on the ComponentProps
-  Shorthands[Primary] &
-    // Add shorthand props for the other slots
+export type ComponentProps<Shorthands extends ObjectShorthandPropsRecord, Primary extends keyof Shorthands = 'root'> =
+  // Allow all props from the primary slot to be specified at the root
+  PropsWithoutRef<Shorthands[Primary]> &
+    // Include shorthand slot props for each of the component's slots, except the primary slot and `root`
     {
-      [Key in Exclude<
-        keyof Shorthands,
-        // Exclude `root` only if it is the primary slot
-        Primary extends 'root' ? Primary : never
-      >]?: ShorthandProps<NonNullable<Shorthands[Key]>>;
-    };
+      [Key in keyof Omit<Shorthands, Primary | 'root'>]?: ShorthandProps<NonNullable<Shorthands[Key]>>;
+    } &
+    // If the primary slot is customized...
+    (Primary extends 'root'
+      ? {}
+      : {
+          // Add a named prop for the primary slot which ONLY accepts className and style
+          [Key in keyof Primary]?: Pick<React.HTMLAttributes<{}>, 'className' | 'style'>;
+        } & {
+          // Add a `root` prop which accepts any native props, but NOT shorthand syntax
+          root?: Shorthands['root'];
+        });
 ```
 
 #### Implementation
 
 Add a function similar to `getNativeElementProps` function, except it handles splitting the `className` and `style` props to the root slot, as well as mixing in the primary slot.
 
-Something along the lines of this (names and API is still a work in progress):
+Something along the lines of this:
 
 ```ts
-const [rootProps, inputProps] = getRootAndPrimaryNativeElementProps(props, 'input');
+const nativeProps = getPartitionedNativeProps(props, 'input');
 const state = {
   // ...
-  root: rootProps,
-  input: inputProps, // primary slot
+  root: nativeProps.root,
+  input: nativeProps.primary, // primary slot
```

Comment:
Technically resolveShorthand for root isn't completely necessary since we don't allow shorthands, but it also handles merging `props.root` and `nativeProps.root`. If you wanted, we could change it to this, but it's a little easier to make mistakes:
`root: { ...nativeProps.root, ...props.root },`